das/state.go

```diff
@@ -1,53 +1,140 @@
 package das
 
-import (
-	"sync"
-	"time"
-)
-
-// state collects information about the DASer process. Currently, there are
-// only two sampling routines: the main sampling routine which performs sampling
-// over current network headers, and the `catchUp` routine which performs sampling
-// over past headers from the last sampled checkpoint.
+// represents current state of sampling
 type state struct {
-	sampleLk sync.RWMutex
-	sample   RoutineState // tracks information related to the main sampling routine
+	priority   []uint64        // list of headers heights that will be sampled with higher priority
+	inProgress map[uint64]bool // keeps track of inProgress items with priority flag stored as value
+	failed     map[uint64]int  // stores heights of failed headers with amount of attempt as value
 
-	catchUpLk sync.RWMutex
-	catchUp   JobInfo // tracks information related to the `catchUp` routine
+	priorityBusy bool // semaphore to allow only one priority item to be processed at time
+
+	maxKnown   uint64 // max known header height
+	next       uint64 // all headers before next were sent to workers
+	minSampled uint64 // all headers before minSampled were successfully sampled
 }
 
-// RoutineState contains important information about the state of a
-// current sampling routine.
-type RoutineState struct {
-	// reports if an error has occurred during the routine's
-	// sampling process
-	Error error `json:"error"`
-	// tracks the latest successfully sampled height of the routine
-	LatestSampledHeight uint64 `json:"latest_sampled_height"`
-	// tracks the square width of the latest successfully sampled
-	// height of the routine
-	LatestSampledSquareWidth uint64 `json:"latest_sampled_square_width"`
-	// tracks whether routine is running
-	IsRunning bool `json:"is_running"`
+func (s *state) handleResult(res result) {
+	if s.inProgress[res.height] {
+		// unblock priority
+		s.priorityBusy = false
+	}
+
+	//TODO: no handling, no retry. just store retry stats for now
+	if res.err != nil {
+		s.failed[res.height]++
+		return
+	}
+
+	delete(s.inProgress, res.height)
+	delete(s.failed, res.height)
+
+	// set minSampled based on min non-failed and done header
+	for s.minSampled < s.next-1 {
+		if _, failed := s.failed[s.minSampled]; failed {
+			break
+		}
+		if _, busy := s.inProgress[s.minSampled]; busy {
+			break
+		}
+		s.minSampled++
+	}
 }
 
-// JobInfo contains information about a catchUp job.
-type JobInfo struct {
-	Start time.Time `json:"start"`
-	End   time.Time `json:"end"`
-	Error error     `json:"error"`
+func (s *state) updateMaxKnown(last uint64) bool {
+	// seen this header before
+	if last == s.maxKnown {
+		return false
+	}
```

Comment:
Did you see this case in practice or do you assume that it might happen? AFAIU, this case is impossible and done on HeaderGetter level, if it does happen, then there is a bug